src/traits.rs

```diff
@@ -1395,6 +1397,199 @@ impl HexDisplay for str {
   }
 }
 
+/// A saturating iterator for usize.
+pub struct SaturatingIterator {
+  count: usize,
+}
+
+impl Iterator for SaturatingIterator {
+  type Item = usize;
+  
+  fn next(&mut self) -> Option<Self::Item> {
+    let old_count = self.count;
+    self.count = self.count.saturating_add(1);
+    Some(old_count)
+  }
+}
+
+/// Abstractions for range-like types.
+pub trait NomRange<Idx> {
+  /// The saturating iterator type.
+  type Iter1: Iterator<Item=Idx>;
+  /// The bounded iterator type.
+  type Iter2: Iterator<Item=Idx>;
+  
+  /// `true` if `item` is contained in the range.
+  fn contains(&self, item: &Idx) -> bool;
+  
+  /// Returns the bounds of this range.
+  fn bounds(&self) -> (Bound<Idx>, Bound<Idx>);
+  
+  /// `true` if the range is inverted.
+  fn is_inverted(&self) -> bool;
+
+  /// Creates a saturating iterator.
+  /// A saturating iterator counts the number of iterations starting from 0 up to the upper bound of this range.
+  /// If the upper bound is infinite the iterator saturates at the largest representable value of its type and
+  /// returns it for all further elements.
+  fn saturating_iter(&self) -> Self::Iter1;
+  
+  /// Creates a bounded iterator.
+  /// A bounded iterator counts the number of iterations starting from 0 up to the upper bound of this range.
+  /// If the upper bounds is infinite the iterator counts up until the amount of iterations has reached the 
+  /// largest representable value of its type and then returns `None` for all further elements.
+  fn bounded_iter(&self) -> Self::Iter2;
+}
+
+impl NomRange<usize> for Range<usize> {
+  type Iter1 = Range<usize>;
+  type Iter2 = Range<usize>;
+  
+  fn bounds(&self) -> (Bound<usize>, Bound<usize>) {(Bound::Included(self.start), Bound::Excluded(self.end))}
+  
+  fn contains(&self, item: &usize) -> bool {RangeBounds::contains(self, item)}
+  
+  fn is_inverted(&self) -> bool {<Range<usize>>::is_empty(self)}
+  
+  fn saturating_iter(&self) -> Self::Iter1 {
+    if self.end == 0 {
+      1..0
+    } else {
+      0..self.end-1
+    }
+  }
+  
+  fn bounded_iter(&self) -> Self::Iter2 {
+    if self.end == 0 {
+      1..0
+    } else {
+      0..self.end-1
+    }
+  }
+}
+
+impl NomRange<usize> for RangeInclusive<usize> {
+  type Iter1 = Range<usize>;
+  type Iter2 = Range<usize>;
+  
+  fn bounds(&self) -> (Bound<usize>, Bound<usize>) {(Bound::Included(*self.start()), Bound::Included(*self.end()))}
+  
+  fn contains(&self, item: &usize) -> bool {RangeBounds::contains(self, item)}
+  
+  fn is_inverted(&self) -> bool {<RangeInclusive<usize>>::is_empty(self)}
```

Comment:
if possible, I'd like to keep nom usable on 1.41.1 for now. As said in the readme, it works on that version without guarantee, but the debian stable upgrade happened quite recently, so people have not migrated quickly to the new version (which has a more recent rustc)

Reply:
Done